### PR TITLE
Added ability to secure DMS assets by user group or role

### LINKED ISSUE #92: Enhancement: Add ability to secure DMS assets by user group or role
As a content author,
I want to be able to secure my DMS documents for certain CMS user groups and/or roles,
so that my documents will be secure from unauthorised users

### Acceptance Criteria

* I can restrict downloading to certain SilverStripe user groups or roles
* If I'm not logged in or not in a given group/role, I can't access a DMS document
* If higher security levels are required then these are assigned at the document level
* The UI in the CMS is easy enough to follow

### PR DESCRIPTION
**Pull request by fspringveldt**
Resolves #92

### CODE REVIEW

**Comment by robbieaverill** [comment]
If there's an alternative to `getSourceAsArray` that is 3.1 compatible that would be cool, but I reckon we could drop 3.1 support if we need to

File: code/model/DMSDocument.php [hunk 1]
```diff
@@ -60,53 +84,56 @@ class DMSDocument extends DataObject implements DMSDocumentInterface
...
 
-    /**
-     * @param Member $member
-     *
-     * @return boolean
-     */
     public function canEdit($member = null)
     {
-        if (!$member || !(is_a($member, 'Member')) || is_numeric($member)) {
+        if (! $member || ! (is_a($member, 'Member')) || is_numeric($member)) {
```
Comment (robbieaverill) [inline code/model/DMSDocument.php]: @fspringveldt what are these changes for?

Comment (fspringveldt) [inline code/model/DMSDocument.php]: @robbieaverill IDE weirdness. Will undo.

File: code/model/DMSDocument.php [hunk 2]
```diff
@@ -67,53 +91,69 @@ class DMSDocument extends DataObject implements DMSDocumentInterface
...
             $member = Member::currentUser();
         }
 
-        // extended access checks
-        $results = $this->extend('canView', $member);
-
-        if ($results && is_array($results)) {
-            if (!min($results)) {
-                return false;
-            }
```
Comment (robbieaverill) [inline code/model/DMSDocument.php]: These extension checks are important - can you add them back in? Without them the checks aren't extensible from user code

File: code/model/DMSDocument.php [hunk 3]
```diff
@@ -123,19 +163,7 @@ public function canEdit($member = null)
...
-            $member = Member::currentUser();
-        }
-
-        $results = $this->extend('canCreate', $member);
-
-        if ($results && is_array($results)) {
-            if (!min($results)) {
-                return false;
-            }
-        }
```
Comment (robbieaverill) [inline code/model/DMSDocument.php]: Again here, add the extension point back

File: code/model/DMSDocument.php [hunk 4]
```diff
@@ -1329,4 +1413,45 @@ protected function getRelatedDocumentsGridField()
...
+     * Checks at least one group is selected if CanViewType || CanEditType == 'OnlyTheseUsers'
+     *
+     * @return ValidationResult
+     */
+    protected function validate()
+    {
+        $valid = parent::validate();
+
+        if ($this->CanViewType == 'OnlyTheseUsers' && !$this->ViewerGroups()->count()) {
+            $valid->error("Selecting 'Only these people' from a viewers list needs at least one group selected.");
```
Comment (robbieaverill) [inline code/model/DMSDocument.php]: Can you add translations for these lines?

File: code/model/DMSDocument.php [hunk 5]
```diff
@@ -445,12 +473,30 @@ public function removeAllTags()
...
+     * point for this was also added.
+     *
+     * To extend use the following from within an Extension subclass:
+     *
+     * <code>
+     *
+     * public function updateGetLink($result){
+     *  //Do something here
+     * }
+     * </code>
```
Comment (sachajudd) [inline code/model/DMSDocument.php]: Could you please fix the indentation to suit our code style eg: // Do something 
and remove the empty line after `<code>`

File: code/model/DMSDocument.php [hunk 6]
```diff
@@ -1095,14 +1141,52 @@ public function getCMSFields()
...
+
+        foreach ($showFields as $name => $extraCss) {
+            $compositeName = "Root.Settings.$name";
+            /** @var FormField $field */
+            if ($field = $settingsFields->fieldByName($compositeName)) {
+                $field->addExtraClass($extraCss);
+                $title = str_replace('page', 'document', $field->Title());
+                $field->setTitle($title);
+
+                //Remove Inherited source option from DropdownField
```
Comment (sachajudd) [inline code/model/DMSDocument.php]: Code style

File: code/model/DMSDocument.php [hunk 7]
```diff
@@ -67,53 +91,69 @@ class DMSDocument extends DataObject implements DMSDocumentInterface
...
-     * @return boolean
-     */
     public function canEdit($member = null)
     {
         if (!$member || !(is_a($member, 'Member')) || is_numeric($member)) {
             $member = Member::currentUser();
         }
 
-        $results = $this->extend('canEdit', $member);
+        //Do early admin check
```
Comment (sachajudd) [inline code/model/DMSDocument.php]: Code style

File: _config.php [hunk 8]
```diff
@@ -15,11 +15,11 @@
...
+    'dms_document_link',
     array('DMSShortcodeHandler', 'handle')
 );
 
 if ($config->get('DMSDocument_versions', 'enable_versions')) {
-	//using the same db relations for the versioned documents, as for the actual documents
-	$config->update('DMSDocument_versions', 'db', $config->get('DMSDocument', 'db'));
+    //using the same db relations for the versioned documents, as for the actual documents
+    $config->update('DMSDocument_versions', 'db', $config->get('DMSDocument', 'db'));
 }
```
Comment (sachajudd) [inline _config.php]: Code style please

File: tests/DMSDocumentTest.php [hunk 9]
```diff
@@ -237,4 +237,103 @@ public function testDocumentHasCmsFieldForManagingRelatedDocuments()
...
+        $this->assertContains('Please log in to view this document', $doc1->getPermissionDeniedReason());
+
+        /** @var DMSDocument $doc2 */
+        $doc2 = $this->objFromFixture('DMSDocument', 'doc-only-these-users');
+        $this->assertContains('You are not authorised to view this document', $doc2->getPermissionDeniedReason());
+
+        /** @var DMSDocument $doc3 */
+        $doc3 = $this->objFromFixture('DMSDocument', 'doc-anyone');
+        $this->assertEquals('', $doc3->getPermissionDeniedReason());
+    }
```
Comment (sachajudd) [inline tests/DMSDocumentTest.php]: Could you please fix the code style in this file

File: code/model/DMSDocument.php (hunk 5 above)
Comment (robbieaverill) [inline code/model/DMSDocument.php]: While you're there, the code example should be PSR-2 style

**Pull request merged**